lib/models/asyncapi.js

```diff
@@ -276,121 +351,139 @@ function addNameToKey(messages, number) {
+      combineArray.forEach(combineSchema => {
+        traverseSchema(combineSchema, cb, null, schemaTypesToIterate);
+      });
+    }
+  };
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.allOfs)) {
     checkCombiningSchemas(schema.allOf());
+  }
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.anyOfs)) {
     checkCombiningSchemas(schema.anyOf());
+  }
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.oneOfs)) {
     checkCombiningSchemas(schema.oneOf());
   }
-
-  seenObj.pop();
 }
 
 /**
- * Go through each channel and for each parameter, and message payload and headers recursively call the callback for each schema.
+ * Go through each channel and for each parameter, and message payload and headers recursively call the cb for each schema.
  * 
  * @private
  * @param {AsyncAPIDocument} doc 
- * @param {FoundSchemaCallback} callback
+ * @param {FoundSchemaCallback} cb
+ * @param {SchemaTypesToIterate[]} schemaTypesToIterate
  */
-function schemaDocument(doc, callback) {
+function traverseAsyncApiDocument(doc, cb, schemaTypesToIterate) {
+  if (!schemaTypesToIterate) {
+    schemaTypesToIterate = [
+      SchemaTypesToIterate.allOfs,
+      SchemaTypesToIterate.anyOfs,
+      SchemaTypesToIterate.arrays,
+      SchemaTypesToIterate.components,
+      SchemaTypesToIterate.headers,
+      SchemaTypesToIterate.objects,
+      SchemaTypesToIterate.oneOfs,
+      SchemaTypesToIterate.parameters,
+      SchemaTypesToIterate.payloads
+    ];
+  }
   if (doc.hasChannels()) {
     doc.channelNames().forEach(channelName => {
       const channel = doc.channel(channelName);
+      traverseChannel(channel, cb, schemaTypesToIterate);
+    });
+  }
+  if (doc.hasComponents() && schemaTypesToIterate.includes(SchemaTypesToIterate.components)) {
+    Object.values(doc.components().schemas()).forEach(s => {
+      traverseSchema(s, cb, null, schemaTypesToIterate);
+    });
+  }
+}
 
-      Object.values(channel.parameters()).forEach(p => {
-        recursiveSchema(p.schema(), callback);
-      });
-
-      if (channel.hasPublish()) {
-        channel.publish().messages().forEach(m => {
-          recursiveSchema(m.headers(), callback);
-          recursiveSchema(m.payload(), callback);
-        });
-      }
-      if (channel.hasSubscribe()) {
-        channel.subscribe().messages().forEach(m => {
-          recursiveSchema(m.headers(), callback);
-          recursiveSchema(m.payload(), callback);
-        });
-      }
+/**
+ * Go through each schema in channel
+ * 
+ * @private
+ * @param {Channel} channel 
+ * @param {FoundSchemaCallback} cb
+ * @param {SchemaTypesToIterate[]} schemaTypesToIterate
+ */
+function traverseChannel(channel, cb, schemaTypesToIterate) {
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.parameters)) {
+    Object.values(channel.parameters()).forEach(p => {
+      traverseSchema(p.schema(), cb, null, schemaTypesToIterate);
+    });
+  }
+  if (channel.hasPublish()) {
```

Comment:
As I remember (but I can be wrong), AsyncAPI only support one publish or subscribe per channel. You cannot combine publish/subscribe. @derberg I am right?

```js
if (channel.hasPublish() || channel.hasSubscribe()) {
    channel.publish().messages().forEach(m => {
      traverseMessage(m, cb, schemaTypesToIterate);
    });
}
```

will be enough.

Reply:
The spec does not specify only one operation is allowed. Also this combination is used in request/reply (since its not support yet) that both gets specified :slightly_smiling_face: